File: common/security/src/main/java/org/yechan/AuthorizeHttpRequestsCustomizer.java
package org.yechan;

import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configurers.AuthorizeHttpRequestsConfigurer;

@FunctionalInterface
public interface AuthorizeHttpRequestsCustomizer {
    void customize(AuthorizeHttpRequestsConfigurer<HttpSecurity>.AuthorizationManagerRequestMatcherRegistry registry);
}

File: common/security/src/main/kotlin/org/yechan/AuthorizeHttpRequestsCustomizer.kt
package org.yechan

import org.springframework.core.Ordered
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configurers.AuthorizeHttpRequestsConfigurer

fun interface AuthorizeHttpRequestsCustomizer {
    fun customize(
        registry: AuthorizeHttpRequestsConfigurer<HttpSecurity>.AuthorizationManagerRequestMatcherRegistry,
    )
}

class PrioritizedAuthorizeHttpRequestsCustomizer(
    private val order: Int,
    private val delegate: AuthorizeHttpRequestsCustomizer,
) : AuthorizeHttpRequestsCustomizer,
    Ordered {
    override fun customize(
        registry: AuthorizeHttpRequestsConfigurer<HttpSecurity>.AuthorizationManagerRequestMatcherRegistry,
    ) {
        delegate.customize(registry)
    }

    override fun getOrder(): Int = order
}
